Add Central Management blacklisting

exekias opened the pull request:
This change allows to define local blacklist for configurations coming
from Central Management. If a configuration block matches the given
regular expession, it will be ignored:

For example:

```
management:
  blacklist:
    output: console
    metricbeat.modules.module: k.{18}s
```

ph commented on x-pack/libbeat/management/blacklist.go:
```diff
@@ -0,0 +1,188 @@
+		return fmt.Errorf("wrong type, expect map")
+	}
+
+	var expand func(key string, value interface{})
```
Not really needed since you are defining it below?

exekias replied:
It's needed because this function is recursive, so internal calls need the definition beforehand

ph commented on x-pack/libbeat/management/blacklist.go:
```diff
@@ -0,0 +1,188 @@
+	return result
+}
+
+func (c *ConfigBlacklist) blacklisted(blockType string, block *api.ConfigBlock) bool {
```
maybe rename to `isBlacklisted()` because its a predicate?

ph commented on x-pack/libbeat/management/blacklist.go:
```diff
@@ -0,0 +1,188 @@
+	}
+
+	return result
+}
```
From what I see, lets say that we receive a config block with a black listed option, we would return log an error and still apply the other config blocks? 

I think to have least surprise as a user it would be better to refuse to apply _all the blocks_, my reasoning is if we instead apply blocks in part we have an inconsistent view of the configuration and this would be unexpected? 

Maybe I am wrong.

ph replied:
Or maybe we don't have the mechanism yet in place to communicate that problem to the user through the UI?

exekias replied:
This is a fair question, I don't know what the user should expect in these cases. So your approach would be to "pause" the beat, and disable all modules & outputs until the configuration is not blacklisted?

We currently have a way to report an error, it refers users to the log file when that happens.

urso replied:
No matter what, we also to signal the UI that configs can not be applied due to the blacklist. Just logging the issue locally, makes it somewhat 'silent'.

ph replied:
Also, people could have access to management but don't have physical access to the machine. 

@urso I think it could be done in two PR, this one to detect errors and another one to improve the feedback loop of a beats since this would certainly require UI changes to support it.

ph commented on x-pack/libbeat/management/blacklist.go:
```diff
@@ -0,0 +1,188 @@
+
+		for _, block := range configs.Blocks {
+			if c.blacklisted(configs.Type, block) {
+				logp.Err("Got a blacklisted configuration, ignoring it")
```
Log which option was blacklisted?

ph commented on x-pack/libbeat/management/blacklist.go:
```diff
@@ -0,0 +1,188 @@
+	return false
+}
+
+func (c *ConfigBlacklist) blockMatches(pattern *regexp.Regexp, segments []string, current *common.Config) bool {
```
Add **is** prefix for predicate.

urso commented on x-pack/libbeat/management/blacklist.go:
```diff
@@ -0,0 +1,188 @@
+		expand(k, val)
+	}
+	return nil
+}
```
What is this type + method supposed to do? The method unpacks the given map into the given map, which is only allocated temporarily? Plus, Unpack as supported by go-ucfg requires a pointer receiver.

The current implementation of the Unpack looks broken to me. Do we have a system test applying a custom blacklisted config?

Why do we need an extra ConfigBlacklistSettings and ConfigBlacklist? One is the settings and the other one is the filter/checker?

How about type names (using full names): `management.BlacklistConfig` and `management.BlacklistCheck`. As we might introduce a whitelist in the future I see a Checker interface and have a package named: `cfgcheck`.

Regarding the Unpack method:

```
type Config struct {  // full name is: blacklist.Config
  patterns map[string]match.Matcher
}

// Note: with this signature go-ucfg will check the config object can be converted to map[string]interface{}
func (s *Config) Unpack(from map[string]interface{}) (err error) {
  s.patterns, err = expandMatchers(from)
  return err
}

func expandMatchers(from map[string]interface{}) (map[string]*matcher.Match, error) {
  settings := flattenConfigs(from)
  patterns := make(map[string]*matcher.Match, len(settings))
  for k, v := range settings {
    m, err := match.Compile(fmt.Sprintf("%s", v))
    if err != nil {
      return nil, errors.Wrapf(err, "failed to compile rule for key '%v'", key)
    }

    patterns[k] = m
  }
  return patterns, nil
}

func flattenConfigs(from map[string]interface{}) map[string]interface{} {
  tmp := map[string]interface{}{}
  for k, v := range from {
    recFlattenConfigs(tmp, k, v)    
  }
  return tmp
}

func recFlattenConfigs(to map[string]interface{}, root string, value interface{}) {
  switch v := value.(type) {
  case map[string]interface{}:
    for key, value := range v {
      recFlattenConfigs(to, fmt.Sprintf("%s.%s", root, key), value)
    }
  case []interface{}:
    for i, value := range v {
      recFlattenConfigs(to, fmt.Sprintf("%s.%d", root, i), value)
    }
  default:
    to[root] = value
  }
}
```

exekias replied:
Thanks for your comments, several things:

- I took the `Unpack` implementation from https://github.com/elastic/beats/blob/master/libbeat/conditions/fields.go#L29-L58, this use case is pretty similar to that
- I need an structure that can be converted back to YAML easily, as the enroll command will end up writing the blacklist, that's why the current one seems convenient.
- It actually works, I'm wondering how since I didn't add a pointer receiver to Unpack, still the map structure is a reference anyway.

urso replied:
> I took the Unpack implementation from /libbeat/conditions/fields.go@master#L29-L58, this use case is pretty similar to that

Not exactly. There are differences.

First of all the unpacker interfaces supported by go-ucfg are: BoolUnpacker, ConfigUnpacker, FloatUnpacker, IntUnpacker, StringUnpacker, UintUnpacker, and Unpacker. We indeed miss an Unpacker interface{} accepting `map[string]interface{}`. So using `interface{}` or `*common.Config` as argument to the method is correct.

When using Unpack, you want to persist changes into the method receiver. Therefore you must use a pointer on the method receiver. I doubt your method is even executed by go-ucfg, but then reflection can be a tricky and surprising at times.

If your method is indeed called (which would be a bug in go-ucfg), it could still work if/because:
- pattern variable is initialized with a `map[string]interface{}` (it must not be `nil`)
- maps are always passed by reference. That is, even if `ConfigBlacklistSettings` is used as value receiver, the underlying map will be copied
- if ConfigBlacklistSettings contains default settings, then these will still be there (you should also clean the map).

Some differences between this implementation and the fields one:
- Fields is a struct, no re-type of map[string]interface{}. This shouldn't be a problem though.
- The signature in the Fields example is actually `func (f *Fields) Unpack(to interface{}) error`. It should say `from` not `to` for the parameter (which is indeed confusing).
- the fields example initializes a local empty map overwriting/removing potential defaults: `f.fields = map[string]interface{}{}`. Your sample code does not remove the initial default settings. In the blacklist case the user settings would be merged with the default settings.
- the expand function in fields.go copies the final result into f.fields (`f.fields[key] = value`). The blacklist code copies the final result back into `m` (`m[key] = fmt.Sprintf("%s", value)`), but does not write any settings into `f`.

Are you sure it actually works as expected? I guess there are 2 potential scenarios why you think it works:
- The tests might be running with the defaults + tests themselves would get you same result no matter if defaults or test settings are applied
- go-ucfg doesn't execute the Unpack method, because `*ConfigBlacklistSettings` does not implement the Unpacker interface -> go-ucfg will detect ConfigBlacklistSettings being a `map[string]interface{}` and just merge all config file settings into ConfigBlacklistSettings as is (no flattening of names).

Tip: Add a fmt.Println(...) to Unpack and check if it's actually executed.

I also noticed your tests are somewhat incomplete. The unit tests call NewConfigBlacklist, but there are no unit-test involving the unpacking of `ConfigBlacklistSettings`. Either create separate unit tests for NewConfigBlacklistSettings or change your current unit tests to use config objects. E.g. change `pattern map[string]string` to `config map[string]interface{}`, and have the test create a config object and unpack the config object into `ConfigBlacklistSettings`.

Is it on purpose that user provided configs are merged with the defaults? Or do we rather want to remove defaults if users provide their own settings?


> - I need an structure that can be converted back to YAML easily, as the enroll command will end up writing the blacklist, that's why the current one seems convenient.

Which is ok I think. The problem is not exectly with the type of ConfigBlacklistSettings (well, a shorter name would be nice :) ), but the implementation of Unpack.

ph replied:
@urso we also have `FlattenedKeys()` in ucfg, It just return the keys not the value, but in this context it could be already used.

urso replied:
Right. We also have a `(MapStr).Flatten` which is similar to `flattenConfigs`.

exekias replied:
Thanks for your input folks, really appreciated, will do the changes

exekias replied:
pushed 0f802c3

urso replied:
Implementation looks good.

urso commented on x-pack/libbeat/management/blacklist.go:
```diff
@@ -0,0 +1,188 @@
+}
+
+// Filter returns a copy of the given ConfigBlocks with the
+func (c *ConfigBlacklist) Filter(configBlocks api.ConfigBlocks) api.ConfigBlocks {
```
I'd prefer to keep the ConfigBlock api a as small as possible. Reading the code, I understand that `ConfigBlock` is be mostly a predicate.

e.g. 

```
type Checker interface {
  Check(string, api.ConfigBlock) bool // return false if the config block is rejected by the checker
}

func FilterBlocks(checker Checker, blocks api.ConfigBlocks) api.ConfigBlocks {
  ...
}

func (c *ConfigBlacklist) Check(type string, blk api.ConfigBlock) bool {
  return !c.isBlacklisted(type, blk)
}
```

We could also wrap the Checker to log/collect all config blocks that do not match for reporting back to kibana.


As all functionality is in the management package, do we even need to export the filter function?

ph replied:
Yes it's better approach.

I would say no, if we encapsulate all the checks (whitelist+blacklist) and we just check if the config is is valid that removes the need do any filter and this would probably simplify things to know what state the beats is currently in.

All config block valid => config applied 
Any invalid block => do not apply the config.

ph replied:
I also thing validate should check all the checks and not fails on the first one so we can report all the errors at once in the UI.

urso replied:
+1 on collecting all errors. Use multierror.

exekias replied:
Not sure I understand the keep `ConfigBlock` API small comment, I don't add anything to ConfigBlock here, as `isBlacklisted` is part of `ConfigBlacklist`.

Do we agree then on treating any a single blacklisted block as a global error that will stop applying new configs? I can do the changes and introduce multierror

ph commented:
@exekias Just ping me if you need another look.

exekias commented:
this should be ready for another look, I added a commit to use multierror and stop applying configs on any blocked config, added a comment here: https://github.com/elastic/beats/pull/9099#discussion_r237106582

ph commented on x-pack/libbeat/management/blacklist.go:
```diff
@@ -0,0 +1,166 @@
+// ConfigBlacklist takes a ConfigBlocks object and filter it based on the given
+// blacklist settings
+type ConfigBlacklist struct {
+	patterns map[string]*regexp.Regexp
```
I will do the change in a followup PR that will include the changelog.
